V3.0 including new tab bar and other user experience changes

mtmcp opened the pull request:
This pull request includes a series of enhancements and bug fixes. The tclmonster fork has added v2.8 and v3.0 branches and this PR will be merged into the tclmonster master branch. Part of this PR is to discuss the possibility of doing similar in this upstream codebase. The intent on the 3.0 version is to update the user experience and that will include dropping support for anything before Tcl/Tk 8.6. Any feedback on the changes welcomed!

A summary of changes is included here:

1. Add a tab bar (at the top) and corresponding object-oriented types to support this and future custom widgets.
2. Update old terminology.
3. Update fonts to support recent fonts across operating systems.
4. Update colors to support dark theme on Windows, macOS, and Linux.
5. Harmonize default proportions (rows, columns) with display aspect ratio.
6. Fix issue with first-tab deletion causing proportions to warp.
7. Add popup menu to edit window.
8. Update find dialog.
9. Add object-oriented widgets.

## Windows:

<img width="1029" height="550" alt="image" src="https://github.com/user-attachments/assets/24398e15-242c-4567-bdd6-9c11db7a096a" />

<img width="990" height="417" alt="Screenshot 2025-09-22 112002" src="https://github.com/user-attachments/assets/59f3d844-2602-422b-b9e7-816940064111" />

## macOS:

<img width="772" height="306" alt="image" src="https://github.com/user-attachments/assets/e49df20b-4e12-45a0-9390-abfee7866c47" />

<img width="709" height="318" alt="image" src="https://github.com/user-attachments/assets/b9012535-038c-4316-abe4-5113881bd163" />

mtmcp committed:
Remove old terminology

mtmcp committed:
Fix sporadic auto-paste, and more.

* Replace old file "exec" with "executable".
* Update fonts to modern macOS/Windows defaults.

mtmcp committed:
Remove redundant font-creation

mtmcp committed:
Merge branch 'rename-vars' into tclmonster-main

mtmcp committed:
Create series of cascading fallback-fonts

mtmcp committed:
Merge branch 'macos-fixes' into tclmonster-main

mtmcp committed:
Add initial tab layout

mtmcp committed:
WIP: adding tab functionality

mtmcp committed:
Implement TabButton object

mtmcp committed:
Add more ui font sizes

mtmcp committed:
Only show tab bar with multiple tabs

mtmcp committed:
Harmonize default UI proportions with screen aspect

mtmcp committed:
Bump version to 3.0

mtmcp committed:
Fix darkmode text default on macOS

mtmcp committed:
Force close button to a square

mtmcp committed:
Keep setgrid when first tab closed

mtmcp committed:
Add tab separators

mtmcp committed:
win32: adjust colors, fix unicode text & radiobutton var.

mtmcp committed:
Set console proportions before setgrid transfer

mtmcp committed:
Attribution for 3.0 enhancements

bohagan1 commented:
Sorry been busy with work. Lots of good stuff. Let me know when your done and I'll merge it. Thanks.

mtmcp commented:
@bohagan1 Sounds good. There are a couple changes I have in mind still, mainly in the status bar area. Should have it all in soon.

mtmcp commented:
Quick update—still working on the fixes—taking things a bit further and got dark-mode working on Windows setting the Window frame with the help of cffi. After trying to work through platform-specific colors I decided to unify the approach and create a simple theme that will support light/dark mode consistently across platforms.

The goal now is a light-weight theme that should satisfy the needs of the console app. I decided to use [Adobe Spectrum](https://spectrum.adobe.com/) design system as a basis (mainly for the colors and basic principles), so the contrast ratios will factor in accessibility and such.

Because this will break what goes in the COLORS array this needs to be merged into 3.0 as a breaking change, before subsequent 3.x releases can be made.

sohamxda7 commented:
looking forward to it

bohagan1 commented:
No rush

mtmcp committed:
Merge branch v3.0-colors

mtmcp committed:
Add FindDialog object

mtmcp committed:
Remove redundant function call

mtmcp committed:
Add docs to about when found

mtmcp committed:
Remove RCS retrieval code

mtmcp committed:
Add about dialog

mtmcp committed:
Open dialogs near focused widget

mtmcp committed:
Open edit windows relative to focused widget

mtmcp committed:
Allow setting custom window

mtmcp committed:
Add popup menu to edit dialog

mtmcp committed:
Replace tabs with spaces

mtmcp commented:
@bohagan1 This PR is ready to be reviewed/merged. I have updated the screenshots to reflect the latest release.

Please note, I removed the RCS retrieval code as it seems unnecessary given how tkcon is distributed these days, and was disabled anyhow. Should the need arise in the future to pull from GitHub in a similar fashion, the logic can of course be searched in the commit history.

In addition, although I mentioned a theme above, I have hardcoded all the hex colours corresponding to the Adobe spectrum palettes, to ensure that TkCon will work with various themes. The palettes are mainly applied to the consoles themselves.

I am open to any suggestions if there are issues or requests re. these changes.

mtmcp commented:
I am starting work on a full Adobe Spectrum theme, however, in case anyone would like to track the progress: [spectrum-tk](https://github.com/tclmonster/spectrum-tk). It will be a bit before it's functional, but nevertheless it will be a useful modern theme taking advantage of SVG and generally targeting Tcl/Tk 8.6/9.x

mtmcp committed:
Forward OPT(title) to sub-interpreters

mtmcp committed:
Clean up alias when Widget destroyed
